graphql_compiler/cost_estimation/analysis.py

```diff
@@ -0,0 +1,238 @@
+# Copyright 2019-present Kensho Technologies, LLC.
+import bisect
+from dataclasses import dataclass
+from typing import Any, Dict, List
+
+from cached_property import cached_property
+
+from ..compiler.compiler_frontend import graphql_to_ir
+from ..compiler.metadata import FilterInfo, QueryMetadataTable
+from ..cost_estimation.cardinality_estimator import estimate_query_result_cardinality
+from ..cost_estimation.int_value_conversion import (
+    convert_int_to_field_value,
+    field_supports_range_reasoning,
+    is_uuid4_type,
+)
+from ..cost_estimation.interval import Interval
+from ..global_utils import PropertyPath, QueryStringWithParameters, VertexPath
+from ..schema import is_meta_field
+from ..schema.schema_info import QueryPlanningSchemaInfo
+from .filter_selectivity_utils import (
+    adjust_counts_for_filters,
+    get_integer_interval_for_filters_on_field,
+)
+
+
+def _convert_int_interval_to_field_value_interval(
+    schema_info: QueryPlanningSchemaInfo, vertex_type: str, field: str, interval: Interval[int]
+) -> Interval[Any]:
+    lower_bound = None
+    upper_bound = None
+    if interval.lower_bound is not None:
+        lower_bound = convert_int_to_field_value(
+            schema_info, vertex_type, field, interval.lower_bound
+        )
+    if interval.upper_bound is not None:
+        upper_bound = convert_int_to_field_value(
+            schema_info, vertex_type, field, interval.upper_bound
+        )
+    return Interval(lower_bound, upper_bound)
+
+
+def get_field_value_intervals(
+    schema_info: QueryPlanningSchemaInfo,
+    query_metadata: QueryMetadataTable,
+    parameters: Dict[str, Any],
+) -> Dict[PropertyPath, Interval[Any]]:
+    """Map each PropertyPath in the query to its field value interval if it can be computed.
+
+    Args:
+        - schema_info: QueryPlanningSchemaInfo
+        - query_metadata: a representation of the query
+        - parameters: the parameters used for the query
+
+    Returns:
+        Dict mapping some PropertyPaths to the interval of filtered values at that field.
+    """
+    field_value_intervals = {}
+    for location, location_info in query_metadata.registered_locations:
+        filter_infos = query_metadata.get_filter_infos(location)
+        vertex_type_name = location_info.type.name
+
+        # Group filters by field
+        single_field_filters: Dict[str, List[FilterInfo]] = {}
+        for filter_info in filter_infos:
+            if len(filter_info.fields) == 0:
+                raise AssertionError(f"Got filter on 0 fields {filter_info} on {vertex_type_name}")
+            elif len(filter_info.fields) == 1:
+                single_field_filters.setdefault(filter_info.fields[0], []).append(filter_info)
+            else:
+                pass  # We don't do anything for multi-field filters yet
```

Comment:
Lets raise `NotImplementedError` instead?

Reply:
We don't raise `NotImplementedError` when the cost estimator is imperfect. We'd have to raise on every query.